crates/sui-types/src/passkey_authenticator.rs

```diff
@@ -0,0 +1,191 @@
+// Copyright (c) 2021, Facebook, Inc. and its affiliates
+// Copyright (c) Mysten Labs, Inc.
+// SPDX-License-Identifier: Apache-2.0
+use crate::crypto::PublicKey;
+use crate::signature_verification::VerifiedDigestCache;
+use crate::{
+    base_types::{EpochId, SuiAddress},
+    crypto::{DefaultHash, Signature, SignatureScheme, SuiSignature},
+    digests::ZKLoginInputsDigest,
+    error::{SuiError, SuiResult},
+    signature::{AuthenticatorTrait, VerifyParams},
+};
+use fastcrypto::hash::{HashFunction, Sha256};
+use fastcrypto::rsa::{Base64UrlUnpadded, Encoding};
+use fastcrypto::secp256r1::{Secp256r1PublicKey, Secp256r1Signature};
+use fastcrypto::traits::VerifyingKey;
+use fastcrypto::{error::FastCryptoError, traits::ToFromBytes};
+use once_cell::sync::OnceCell;
+use passkey::types::webauthn::CollectedClientData;
+use schemars::JsonSchema;
+use serde::{Deserialize, Serialize};
+use shared_crypto::intent::IntentMessage;
+use std::hash::Hash;
+use std::hash::Hasher;
+use std::sync::Arc;
+
+#[cfg(test)]
+#[path = "unit_tests/passkey_authenticator_test.rs"]
+mod passkey_authenticator_test;
+
+/// An zk login authenticator with all the necessary fields.
+#[derive(Debug, Clone, JsonSchema, Serialize, Deserialize)]
+#[serde(rename_all = "camelCase")]
+pub struct PasskeyAuthenticator {
+    authenticator_data: Vec<u8>, // should size limit be enforced?
+    client_data: Vec<u8>,        // should size limit be enforced?
+    user_signature: Signature,
+    #[serde(skip)]
+    pub bytes: OnceCell<Vec<u8>>,
+}
+
+impl PasskeyAuthenticator {
+    // test only
+    pub fn new(
+        authenticator_data: Vec<u8>,
+        client_data: Vec<u8>,
+        user_signature: Signature,
+    ) -> Self {
+        Self {
+            authenticator_data,
+            client_data,
+            user_signature,
+            bytes: OnceCell::new(),
+        }
+    }
+
+    pub fn get_pk(&self) -> SuiResult<PublicKey> {
+        PublicKey::try_from_bytes(
+            SignatureScheme::Secp256r1,
+            self.user_signature.public_key_bytes(),
+        )
+        .map_err(|_| SuiError::InvalidAuthenticator)
+    }
```

Comment:
If we only support Secp256r1 then lets just enforce that at the type level. So don't use `Signature` but Secp256r1Signature instead.

Reply:
the Signature contains is a compact bytes representation of flag || sig || pk in bytes. we could consider use struct { pk, sig } but it will be a new serialization that ts needs to enshrine as well.

Reply:
We can still have it serialized as flag || sig || pk, but it should be enforced at deser time that it is the correct type